Realm/Realm.Sync/Subscription.cs

```diff
@@ -0,0 +1,175 @@
+        /// </summary>
+        /// <value>The state of the subscription.</value>
+        public SubscriptionState State { get; private set; }
+
+        /// <summary>
+        /// Gets a value indicating what error (if any) has occurred while processing the subscription.
+        /// If the <see cref="State"/> is not <see cref="SubscriptionState.Error"/>, this will be <c>null</c>.
+        /// </summary>
+        /// <value>An instance of <see cref="Exception"/> if an error has occurred; <c>null</c> otherwise.</value>
+        public Exception Error { get; private set; }
+
+        /// <summary>
+        /// Gets the query that this subscription is associated with. Regardless of the state of the subscription,
+        /// this value will reflect the results in the local Realm. This allows you to databind to this property
+        /// immediately and show the last synchronized data. If the <see cref="State"/> is <see cref="SubscriptionState.Invalidated"/>,
+        /// the values returned will not be an adequate representation of the state of the remote Realm.
+        /// </summary>
+        /// <value>
+        /// A queryable collection that can be further filtered, ordered, or observed for changes.
+        /// </value>
+        public IQueryable<T> Results { get; }
+
+        internal Subscription(SubscriptionHandle handle, RealmResults<T> query)
+        {
+            Results = query;
+
+            _handle = handle;
+            State = _handle.GetState();
+
+            var managedSubscriptionHandle = GCHandle.Alloc(this, GCHandleType.Weak);
+            _subscriptionToken = _handle.AddNotificationCallback(GCHandle.ToIntPtr(managedSubscriptionHandle), SubscriptionCallback);
+        }
+
+        /// <summary>
+        /// Waits for the subscription to complete synchronizing (equivalent to transitioning to the
+        /// <see cref="SubscriptionState.Complete"/> state.
+        /// </summary>
+        /// <returns>
+        /// An awaitable task, that, upon completion, indicates that the objects matching the specified query
+        /// have been synchronized to the local Realm.
+        /// </returns>
+        public Task WaitForSynchronizationAsync()
+        {
+            return _syncTcs.Task;
+        }
+
+        internal async Task UnsubscribeAsync()
+        {
+            _handle.Unsubscribe();
+
+            var tcs = new TaskCompletionSource<object>();
+            PropertyChangedEventHandler handler = null;
+            handler = new PropertyChangedEventHandler((s, e) =>
+            {
+                switch (State)
+                {
+                    case SubscriptionState.Invalidated:
+                        tcs.TrySetResult(null);
+                        break;
+                    case SubscriptionState.Error:
+                        tcs.TrySetException(Error);
+                        break;
+                }
+            });
+
+            PropertyChanged += handler;
+            await tcs.Task;
```

Comment:
If the task is faulted the following two lines will never execute. Maybe `try/finally`?